src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs

```diff
@@ -47,14 +47,14 @@ internal ServiceProvider(ICollection<ServiceDescriptor> serviceDescriptors, Serv
             Root = new ServiceProviderEngineScope(this, isRootScope: true);
             _engine = GetEngine();
             _createServiceAccessor = CreateServiceAccessor;
-            _realizedServices = new ConcurrentDictionary<Type, Func<ServiceProviderEngineScope, object?>>();
+            _realizedServices = new ConcurrentDictionary<ServiceIdentifier, Func<ServiceProviderEngineScope, object?>>();
 
             CallSiteFactory = new CallSiteFactory(serviceDescriptors);
             // The list of built in services that aren't part of the list of service descriptors
             // keep this in sync with CallSiteFactory.IsService
-            CallSiteFactory.Add(typeof(IServiceProvider), new ServiceProviderCallSite());
-            CallSiteFactory.Add(typeof(IServiceScopeFactory), new ConstantCallSite(typeof(IServiceScopeFactory), Root));
-            CallSiteFactory.Add(typeof(IServiceProviderIsService), new ConstantCallSite(typeof(IServiceProviderIsService), CallSiteFactory));
+            CallSiteFactory.Add(ServiceIdentifier.FromServiceType(typeof(IServiceProvider)), new ServiceProviderCallSite());
+            CallSiteFactory.Add(ServiceIdentifier.FromServiceType(typeof(IServiceScopeFactory)), new ConstantCallSite(typeof(IServiceScopeFactory), Root));
+            CallSiteFactory.Add(ServiceIdentifier.FromServiceType(typeof(IServiceProviderIsService)), new ConstantCallSite(typeof(IServiceProviderIsService), CallSiteFactory));
```

Comment:
```suggestion
            CallSiteFactory.Add(ServiceIdentifier.FromServiceType(typeof(IServiceProviderIsService)), new ConstantCallSite(typeof(IServiceProviderIsService), CallSiteFactory));
            CallSiteFactory.Add(ServiceIdentifier.FromServiceType(typeof(IKeyedServiceProviderIsService)), new ConstantCallSite(typeof(IServiceProviderIsService), CallSiteFactory));
```

I now feel way more strongly that `IKeyedServiceProviderIsService` should implement `IServiceProviderIsService`. Once things like RDF and RDG start supporting `[FromKeyedServices("name")]` parameters, it's probably going to want to be able to check for the existence of both keyed and non-keyed services so bad configurations can fail at startup, and needing to resolve two different services is annoying. @captainsafia 

I also worry a bit about the discoverability of `IKeyedServiceProviderIsService` because no one is going to interact with `IKeyedServiceProvider` directly, only through the `GetKeyedService` extension method on `IServiceProvider`. I wonder if `IServiceProviderIsKeyedService` is more discoverable since it also starts with `IServiceProvider`. Granted `IServiceProviderIsService` is a terrible name to start from, and maybe we don't need it to be super discoverable.

I feel more strongly that we should at least have `IKeyedServiceProviderIsService : IServiceProviderIsService`. I cannot imagine a container supporting detecting only keyed services.